roachtest: add new panic mutator to upreplicate before panic

dev-kyle opened the pull request:
This change adds a new mutator that is similar to the panic node mutator, but will also upreplicate to an RF of 5 before panicking, and drop back down to 3 after recovering. This allows for a much larger duration of leaving the node down before recovering, as the cluster can now survive 2 nodes being down. The current implementation is only limited to tests that already have 5 nodes in the cluster, so this mutator remains disabled by default.

Epic: none

Release note: none

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150853)
<!-- Reviewable:end -->

dev-kyle committed:
roachtest: add more expansive panic node mutator

This change adds a new mutator that is similar to the panic node
mutator, but will also upreplicate to an RF of 5 before panicking, and
drop back down to 3 after recovering. This allows for a much larger
duration of leaving the node down before recovering, as the cluster can
now survive 2 nodes being down. The current implementation is only
limited to tests that already have 5 nodes in the cluster, so this
mutator remains disabled by default.

Epic: none

Release note: none

DarrylWong commented on pkg/cmd/roachtest/tests/tpcc.go:
```diff
@@ -568,20 +577,25 @@ func runTPCCMixedHeadroom(ctx context.Context, t test.Test, c cluster.Cluster) {
 	// The full 6.5m import ran into out of disk errors (on 250gb machines),
 	// hence division by two.
 	bankRows := 65104166 / 2
+	if minimizedImport {
```
Do we need both changing the `rebalance` max rate and reducing the import size? Can we get away with just one of the two?

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -652,3 +652,118 @@ func selectPartitions(
+	return 0
+}
+
+func (m replicatePanicMutator) Generate(
```
Should this be implemented as a different flavor of the regular `PanicNodeMutator`? That way they are mutually exclusive and you might be able to reuse some of the logic?

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/steps.go:
```diff
@@ -969,3 +969,74 @@ func (s networkPartitionRecoveryStep) Run(
 }
+
+type alterReplicationFactorStep struct {
+	rf       int
```
nit: replicationFactor

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/steps.go:
```diff
@@ -969,3 +969,74 @@ func (s networkPartitionRecoveryStep) Run(
+
+type alterReplicationFactorStep struct {
+	rf       int
+	max_rate string
```
nit: snapshotRebalanceMaxRate

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/steps.go:
```diff
@@ -969,3 +969,74 @@ func (s networkPartitionRecoveryStep) Run(
+	ticker := time.NewTicker(pollInterval)
+	defer ticker.Stop()
+
+	stmt = fmt.Sprintf("SELECT count(1) FROM crdb_internal.ranges WHERE array_length(replicas, 1) < %d", s.rf)
```
Can we not use the `WaitForReplicationFactor` helper in `roachtestutil` here?

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -409,6 +409,10 @@ func (m panicNodeMutator) Generate(
 
+	// If we have at least 5 nodes, we can safely upreplicate to 5X before panicking a node.
+	// This allows for a longer panic duration before recovery.
+	isLargeCluster := len(nodeList) >= 5
```
nit: `supportsUpreplication`?

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -426,13 +430,18 @@ func (m panicNodeMutator) Generate(
-			_, restart := s.impl.(restartWithNewBinaryStep)
+			// cause the cluster to lose quorum, so we avoid any system restarts. If
+			// the cluster has a high enough node count however, we can upreplicate
+			// to 5X before panicking, allowing us to safely restart nodes.
```
>allowing us to safely restart nodes

nit: safely restart _other_ nodes?

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -426,13 +430,18 @@ func (m panicNodeMutator) Generate(
+			// to 5X before panicking, allowing us to safely restart nodes.
+			restartImpl, restart := s.impl.(restartWithNewBinaryStep)
+			if isLargeCluster {
+				restart = restart && restartImpl.node == targetNode[0]
```
I think I know what you are trying to do here, but lets explain it.

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -445,7 +454,7 @@ func (m panicNodeMutator) Generate(
 			}
 
-			return restart || waitForStable || runHook || s.context.System.hasUnavailableNodes
+			return restart || waitForStable || (runHook && !planner.options.hooksSupportFailureInjection) || s.context.System.hasUnavailableNodes
```
We should make this change for the networkpartion mutator too.

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/steps.go:
```diff
@@ -969,3 +971,53 @@ func (s networkPartitionRecoveryStep) Run(
+	); err != nil {
+		return errors.Wrap(err, "failed to change replication factor on system")
+	}
+	if h.IsMultitenant() {
```
Do we need to change the RF for separate process tenants? I don't think so but I could be wrong?

Dev-Kyle replied:
Forgot to remove this from when we were debugging the long replication time, should not need to change RF for separate process tenants - removed

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/steps.go:
```diff
@@ -969,3 +971,43 @@ func (s networkPartitionRecoveryStep) Run(
+		return errors.Wrapf(err, "failed to create logger %s", loggerName)
+	}
+
+	l.Printf("waiting to reach replication factor of %dX", s.replicationFactor)
```
nit: `l.Printf("waiting to reach replication factor of %dX; details in %s.log`

DarrylWong commented on pkg/cmd/roachtest/roachtestutil/mixedversion/mutators.go:
```diff
@@ -426,13 +430,21 @@ func (m panicNodeMutator) Generate(
+			restartImpl, restart := s.impl.(restartWithNewBinaryStep)
+			if supportsUpReplication {
+				// We can restart other nodes, but we do not want to restart the
+				// node that is being panicked, as this can cause conflict with the
```
nit: `as the panic recover step expects the node to be down in order to restart it`. Just to iterate what the actual issue is and to clarify its a simplification we make in the framework rather than an actual crdb limitation.

DarrylWong reviewed (commented):
Let me take it for a test run first, but otherwise LGTM.
